ign_ros2_control/src/ign_system.cpp

```diff
@@ -580,42 +776,133 @@ IgnitionSystem::perform_command_mode_switch(
+  const rclcpp::Duration & period)
 {
+  // refresh params
+  param_listener_->refresh_dynamic_parameters();
+  params_ = param_listener_->get_params();
+
   for (unsigned int i = 0; i < this->dataPtr->joints_.size(); ++i) {
+
+    // assuming every joint has axis
+    const auto * jointAxis =
+      this->dataPtr->ecm->Component<ignition::gazebo::components::JointAxis>(
+      this->dataPtr->joints_[i].sim_joint);
+
+    // update PIDs
+    this->dataPtr->joints_[i].pid_pos.SetPGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].p_pos);
+    this->dataPtr->joints_[i].pid_pos.SetIGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].i_pos);
+    this->dataPtr->joints_[i].pid_pos.SetDGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].d_pos);
+    this->dataPtr->joints_[i].pid_pos.SetIMax(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].iMax_pos);
+    this->dataPtr->joints_[i].pid_pos.SetIMin(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].iMin_pos);
+    this->dataPtr->joints_[i].pid_pos.SetCmdMax(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdMax_pos);
+    this->dataPtr->joints_[i].pid_pos.SetCmdMin(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdMin_pos);
+    this->dataPtr->joints_[i].pid_pos.SetCmdOffset(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdOffset_pos);
+
+    this->dataPtr->joints_[i].pid_vel.SetPGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].p_vel);
+    this->dataPtr->joints_[i].pid_vel.SetIGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].i_vel);
+    this->dataPtr->joints_[i].pid_vel.SetDGain(
+      params_.gains.joints_map[this->dataPtr->joints_[i].name].d_vel);
+    this->dataPtr->joints_[i].pid_vel.SetIMax(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].iMax_vel);
+    this->dataPtr->joints_[i].pid_vel.SetIMin(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].iMin_vel);
+    this->dataPtr->joints_[i].pid_vel.SetCmdMax(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdMax_vel);
+    this->dataPtr->joints_[i].pid_vel.SetCmdMin(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdMin_vel);
+    this->dataPtr->joints_[i].pid_vel.SetCmdOffset(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdOffset_vel);
+
     if (this->dataPtr->joints_[i].joint_control_method & VELOCITY) {
-      if (!this->dataPtr->ecm->Component<ignition::gazebo::components::JointVelocityCmd>(
-          this->dataPtr->joints_[i].sim_joint))
-      {
+
+      double velocity_error = this->dataPtr->joints_[i].joint_velocity - std::clamp(
+        this->dataPtr->joints_[i].joint_velocity_cmd, -1.0 * jointAxis->Data().MaxVelocity(),
+        jointAxis->Data().MaxVelocity());
```

Comment:
Maybe this deserves to be split up into a couple of lines--it's hard to read as-is.

Reply:
[e40142c](https://github.com/ros-controls/gz_ros2_control/pull/122/commits/e40142c66370bbab409963333b429863e6dfabe1)